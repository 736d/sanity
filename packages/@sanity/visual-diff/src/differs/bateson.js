/* eslint-disable complexity */
/* eslint-disable max-depth, id-length */
import difference from 'lodash/difference'
import intersection from 'lodash/intersection'
import flattenDeep from 'lodash/flattenDeep'

const defaultOptions = {
  ignoreFields: ['_id', '_updatedAt', '_createdAt', '_rev', '_weak']
}

function typeOf(value) {
  if (Array.isArray(value)) return 'array'
  if (typeof value === 'object') return 'object'
  if (value === null || value === undefined) return 'null'
  return typeof value
}

export function isSameType(a, b) {
  if (a === null && b === null) return true
  if (a === null && b !== null) return false
  if (a !== null && b === null) return false
  if (typeof a !== typeof b) return false
  if (typeof a === 'object') {
    if (a._type !== b._type) return false
  }
  return true
}

function sanityType(value) {
  if (value._type) return value._type
  return typeOf(value)
}

function resolveSummarizerForType(a, b, type, path, summarizers) {
  const summarizerForTypeOperation = summarizers[type]
  return summarizerForTypeOperation ? summarizerForTypeOperation.resolve(a, b, path) : null
}

function accumulateChangeSummaries(a, b, path, options) {
  if (!isSameType(a, b)) {
    return [{operation: 'replace', from: a, to: b}]
  }

  const typeWeAreOperatingOn = typeOf(a) // We can use this, as a and b are guaranteed to be the same type

  switch (typeWeAreOperatingOn) {
    case 'object': {
      const result = []

      const summarizerForType = resolveSummarizerForType(
        a,
        b,
        sanityType(a),
        path,
        options.summarizers
      )

      let ignoreFields = options.ignoreFields || []

      if (summarizerForType && summarizerForType.changes) {
        ignoreFields = summarizerForType.fields.concat(ignoreFields)
        if (summarizerForType.fields.length === 0) {
          // An empty fields array means the whole thing has been handled
          return summarizerForType.changes
        }
        result.concat(
          summarizerForType.changes.map(change => {
            change.path = path
            return change
          })
        )
      }

      const [aFields, bFields] = [
        difference(Object.keys(a), ignoreFields),
        difference(Object.keys(b), ignoreFields)
      ]

      const removed = difference(aFields, bFields)
      removed.forEach(field => {
        result.push({operation: 'remove', path, from: a[field]})
      })

      const added = difference(bFields, aFields)
      added.forEach(field => {
        result.push({operation: 'add', path, to: b[field]})
      })

      const kept = intersection(aFields, bFields)
      kept.forEach(field => {
        const fieldA = a[field]
        const fieldB = b[field]
        const changes = accumulateChangeSummaries(fieldA, fieldB, path.concat(field), options)
        if (changes.length > 0) {
          result.push(changes)
        }
      })

      return result
    }

    case 'array': {
      const result = []
      const aElements = {}
      const bElements = {}

      a.forEach(element => {
        aElements[element._key] = element
      })
      b.forEach(element => {
        bElements[element._key] = element
      })
      const [aKeys, bKeys] = [Object.keys(aElements), Object.keys(bElements)]

      intersection(aKeys, bKeys).forEach(key => {
        const elementA = aElements[key]
        const elementB = bElements[key]
        const changes = accumulateChangeSummaries(elementA, elementB, path.concat(key), options)
        if (changes.length > 0) {
          result.push(changes)
        }
      })

      // something has been added
      difference(bKeys, aKeys).forEach(key => {
        result.push({operation: 'add', key, path, to: bElements[key]})
      })

      // something has been removed
      difference(aKeys, bKeys).forEach(key => {
        result.push({operation: 'remove', key, path, from: aElements[key]})
      })

      return result
    }

    default:
      if (a !== b) {
        const summarizerForType = resolveSummarizerForType(
          a,
          b,
          sanityType(a),
          path,
          options.summarizers
        )

        if (summarizerForType && summarizerForType.changes) {
          return summarizerForType.changes.map(change => {
            change.path = path
            return change
          })
        }

        return [{operation: 'edit', path, from: a, to: b}]
      }
      return []
  }
}

export default function changeSummaries(a, b, opts = {}) {
  const options = {...defaultOptions, ...opts}
  const nestedSummaries = accumulateChangeSummaries(a, b, [], options)
  return flattenDeep(nestedSummaries)
}
