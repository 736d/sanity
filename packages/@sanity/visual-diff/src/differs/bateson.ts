/* eslint-disable complexity */
/* eslint-disable max-depth, id-length */
import {BatesonOptions, Operation} from '../@types/visual-diff'

import difference from 'lodash/difference'
import intersection from 'lodash/intersection'
import flattenDeep from 'lodash/flattenDeep'

const defaultOptions: BatesonOptions = {
  ignoreFields: ['_id', '_updatedAt', '_createdAt', '_rev', '_weak']
}

function typeOf(value: any): string {
  if (Array.isArray(value)) return 'array'
  if (typeof value === 'object') return 'object'
  if (value === null || value === undefined) return 'null'
  return typeof value
}

export function isSameType(a: any, b: any): boolean {
  if (a === null && b === null) return true
  if (a === null && b !== null) return false
  if (a !== null && b === null) return false
  if (typeof a !== typeof b) return false
  if (typeof a === 'object' && a._type !== b._type) return false
  return true
}

function sanityType(value: any): string {
  if (value._type) return value._type
  return typeOf(value)
}

function accumulateChangeSummaries(a: any, b: any, path: string[], options: BatesonOptions): Operation[] {
  if (!isSameType(a, b)) {
    return [{operation: 'replace', path, from: a, to: b}]
  }

  const typeWeAreOperatingOn = typeOf(a) // We can use this, as a and b are guaranteed to be the same type
  switch (typeWeAreOperatingOn) {
    case 'object': {
      const result: Operation[] = []
      const summarizerForTypeOperation = options.summarizers[sanityType(a)]
      const summary = summarizerForTypeOperation
        ? summarizerForTypeOperation.resolve(a, b, path)
        : null

      let ignoreFields = options.ignoreFields || []

      if (summary && summary.changes) {
        if (summary.fields.length === 0) {
          // An empty fields array means the whole thing has been handled
          return summary.changes.map(change => ({
            ...change,
            path
          }))
        }
        ignoreFields = summary.fields.concat(ignoreFields)

        summary.changes.forEach(change => {
          result.push({
            ...change,
            path
          })
        })
      }

      const [aFields, bFields] = [
        difference(Object.keys(a), ignoreFields),
        difference(Object.keys(b), ignoreFields)
      ]

      const removed = difference(aFields, bFields)
      removed.forEach(field => {
        result.push({operation: 'remove', path, from: a[field]})
      })

      const added = difference(bFields, aFields)
      added.forEach(field => {
        result.push({operation: 'add', path, to: b[field]})
      })

      const kept = intersection(aFields, bFields)
      kept.forEach(field => {
        const fieldA = a[field]
        const fieldB = b[field]
        const changes = accumulateChangeSummaries(fieldA, fieldB, path.concat(field), options)
        if (changes.length > 0) {
          changes.forEach(change => result.push(change))
        }
      })

      return result
    }

    case 'array': {
      const result: Operation[] = []
      const aElements = {}
      const bElements = {}

      a.forEach(element => {
        aElements[element._key] = element
      })
      b.forEach(element => {
        bElements[element._key] = element
      })
      const [aKeys, bKeys] = [Object.keys(aElements), Object.keys(bElements)]

      intersection(aKeys, bKeys).forEach(key => {
        const elementA = aElements[key]
        const elementB = bElements[key]
        const changes = accumulateChangeSummaries(elementA, elementB, path.concat(key), options)
        if (changes.length > 0) {
          changes.forEach(change => result.push(change))
        }
      })

      // something has been added
      difference(bKeys, aKeys).forEach(key => {
        result.push({operation: 'add', path: path.concat({_key: key}), to: bElements[key]})
      })

      // something has been removed
      difference(aKeys, bKeys).forEach(key => {
        result.push({operation: 'remove', path: path.concat({_key: key}), from: aElements[key]})
      })

      return result
    }

    default:
      if (a !== b) {
        const summarizerForTypeOperation = options.summarizers[sanityType(a)]
        const summary = summarizerForTypeOperation
          ? summarizerForTypeOperation.resolve(a, b, path)
          : null

        if (summary && summary.changes) {
          return summary.changes.map(change => ({
            ...change,
            path
          }))
        }

        return [{operation: 'edit', path, from: a, to: b}]
      }

      return []
  }
}

export default function changeSummaries(a: any, b: any, opts: BatesonOptions = {}) {
  const options = {...defaultOptions, ...opts}
  const nestedSummaries = accumulateChangeSummaries(a, b, [], options)
  return flattenDeep(nestedSummaries)
}
