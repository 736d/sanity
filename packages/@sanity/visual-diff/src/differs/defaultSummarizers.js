/* eslint-disable id-length */

// If the differ takes responsibility for the change in question
// --> return an array of summaries, empty array means no summary needed
// --> return null to "drop the ball" and rely on a default summary to be created

function extractText(blockContent) {
  return blockContent.children
    .map(item => (item._type == 'span' ? item.text : null))
    .filter(Boolean)
    .join('')
}

const differs = {
  block: {
    resolve: (a, b) => {
      const aText = extractText(a)
      const bText = extractText(b)
      if (aText !== bText) {
        return [
          {
            op: 'editText',
            type: 'block',
            from: aText,
            to: bText
          }
        ]
      }
      return []
    }
  },
  string: {
    resolve: (a, b) => {
      return [
        {
          op: 'editText',
          type: 'string',
          from: a,
          to: b
        }
      ]
    }
  },
  image: {
    resolve: (a, b) => {
      if (!a.asset && b.asset) {
        return [{op: 'addImage', field: 'asset', value: b.asset._ref}]
      } else if (a.asset && !b.asset) {
        return [{op: 'removeImage', from: a.asset._ref}]
      } else if (a.asset && b.asset && a.asset._ref !== b.asset._ref) {
        return [{op: 'replaceImage', from: a.asset._ref, to: b.asset._ref}]
      }
      return null
    }
  }
}

export default differs
