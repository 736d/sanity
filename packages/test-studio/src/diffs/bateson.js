/* eslint-disable max-depth, id-length */
import {difference, intersection} from 'lodash'

const context = {
  differs: {
    block: (a, b) => {
      const aText = extractText(a.children)
      const bText = extractText(b.children)
      if (aText !== bText) {
        return [
          {
            op: 'textEdit',
            type: 'block',
            from: aText,
            to: bText
          }
        ]
      }
      return []
    },
    string: (a, b) => {
      return [
        {
          op: 'textEdit',
          type: 'string',
          from: a,
          to: b
        }
      ]
    },
    image: (a, b) => {
      if (a.asset && b.asset && a.asset._ref !== b.asset._ref) {
        return [{op: 'replaceImage', from: a.asset._ref, to: b.asset._ref}]
      }
      return []
    }
  },
  ignore: ['_updatedAt', '_createdAt', '_rev', '_weak']
}

function typeOf(value) {
  if (Array.isArray(value)) return 'array'
  if (typeof value === 'object') return 'object'
  if (value === null || value === undefined) return 'null'
  return typeof value
}

function isSameType(a, b) {
  if (typeof a !== typeof b) return false
  if (Array.isArray(a) && Array.isArray(b)) return true
  if (a === null && b === null) return true
  if (typeof a === 'object') {
    if (a._type !== b._type) return false
  }
  return true
}

function sanityType(value) {
  if (value._type) return value._type
  return typeOf(value)
}

function diff(ctx, path, a, b) {
  if (!isSameType(a, b)) {
    return [{op: 'replace', from: a, to: b}]
  }

  switch (typeOf(a)) {
    case 'object': {
      const result = []
      if (context.differs[a._type]) {
        return context.differs[a._type](a, b)
        result.push(...context.differs[a._type](a, b))
      }
      const [aFields, bFields] = [
        difference(Object.keys(a), context.ignore || []),
        difference(Object.keys(b), context.ignore || [])
      ]
      const removed = difference(aFields, bFields)
      removed.forEach(field => {
        result.push({op: 'remove', field})
      })
      const added = difference(bFields, aFields)
      added.forEach(field => {
        result.push({op: 'set', field, value: b[field]})
      })
      const kept = intersection(aFields, bFields)
      kept.forEach(field => {
        const fieldA = a[field]
        const fieldB = b[field]
        const changes = diff(ctx, path.concat(field), fieldA, fieldB)
        const type = isSameType(fieldA, fieldB) ? sanityType(fieldA) : null
        if (changes.length > 0) {
          result.push({op: 'modifyField', type, field, changes})
        }
      })
      return result
    }

    case 'array': {
      const result = []
      const aElements = {}
      const bElements = {}
      a.forEach(element => {
        aElements[element._key] = element
      })
      b.forEach(element => {
        bElements[element._key] = element
      })
      const [aKeys, bKeys] = [Object.keys(aElements), Object.keys(bElements)]

      intersection(aKeys, bKeys).forEach(key => {
        const elementA = aElements[key]
        const elementB = bElements[key]
        const changes = diff(ctx, path.concat(key), elementA, elementB)
        const type = isSameType(elementA, elementB) ? sanityType(elementA) : null
        if (changes.length > 0) {
          result.push({op: 'modifyEntry', type, key, changes})
        }
      })
      difference(bKeys, aKeys).forEach(key => {
        result.push({op: 'remove', key})
      })
      return result
    }

    default:
      if (a !== b) {
        if (context.differs[typeOf(a)]) {
          return context.differs[typeOf(a)](a, b)
        }
        return [{op: 'edit', from: a, to: b}]
      }
      return []
  }
}

function extractText(blockContent) {
  return blockContent
    .map(item => (item._type == 'span' ? item.text : null))
    .filter(Boolean)
    .join(' ')
}

export default function bateson(a, b) {
  return diff(context, [], a, b)
}
