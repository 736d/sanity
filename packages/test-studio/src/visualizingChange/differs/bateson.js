/* eslint-disable complexity */
/* eslint-disable max-depth, id-length */
import {difference, intersection} from 'lodash'

const defaultContext = {
  ignore: ['_id', '_updatedAt', '_createdAt', '_rev', '_weak']
}

function typeOf(value) {
  if (Array.isArray(value)) return 'array'
  if (typeof value === 'object') return 'object'
  if (value === null || value === undefined) return 'null'
  return typeof value
}

function isSameType(a, b) {
  if (typeof a !== typeof b) return false
  if (Array.isArray(a) && Array.isArray(b)) return true
  if (a === null && b === null) return true
  if (typeof a === 'object') {
    if (a._type !== b._type) return false
  }
  return true
}

function sanityType(value) {
  if (value._type) return value._type
  return typeOf(value)
}

function diff(context, path, a, b) {
  if (!isSameType(a, b)) {
    return [{op: 'replace', from: a, to: b}]
  }

  switch (typeOf(a)) {
    case 'object': {
      const result = []

      const differForType = context.differs[a._type]
      if (differForType) {
        // There might be a differ for this type, but maybe not for every operation imaginable
        const summary = differForType(a, b)
        if (summary) {
          return summary
        }
      }
      const [aFields, bFields] = [
        difference(Object.keys(a), context.ignore || []),
        difference(Object.keys(b), context.ignore || [])
      ]
      const removed = difference(aFields, bFields)
      removed.forEach(field => {
        result.push({op: 'remove', field})
      })
      const added = difference(bFields, aFields)
      added.forEach(field => {
        result.push({op: 'set', field, value: b[field]})
      })
      const kept = intersection(aFields, bFields)
      kept.forEach(field => {
        const fieldA = a[field]
        const fieldB = b[field]
        const changes = diff(context, path.concat(field), fieldA, fieldB)
        const type = isSameType(fieldA, fieldB) ? sanityType(fieldA) : null
        if (changes.length > 0) {
          result.push({op: 'modifyField', type, field, changes})
        }
      })
      return result
    }

    case 'array': {
      const result = []
      const aElements = {}
      const bElements = {}
      a.forEach(element => {
        aElements[element._key] = element
      })
      b.forEach(element => {
        bElements[element._key] = element
      })
      const [aKeys, bKeys] = [Object.keys(aElements), Object.keys(bElements)]

      intersection(aKeys, bKeys).forEach(key => {
        const elementA = aElements[key]
        const elementB = bElements[key]
        const changes = diff(context, path.concat(key), elementA, elementB)
        const type = isSameType(elementA, elementB) ? sanityType(elementA) : null
        if (changes.length > 0) {
          result.push({op: 'modifyEntry', type, key, changes})
        }
      })
      difference(bKeys, aKeys).forEach(key => {
        result.push({op: 'remove', key})
      })
      return result
    }

    default:
      if (a !== b) {
        const differForType = context.differs[typeOf(a)]
        if (differForType) {
          const summary = differForType(a, b)
          if (summary) {
            return summary
          }
        }
        return [{op: 'edit', from: a, to: b}]
      }
      return []
  }
}

export default function bateson(a, b, options = {}) {
  const context = {...defaultContext, ...options}
  return diff(context, [], a, b)
}
