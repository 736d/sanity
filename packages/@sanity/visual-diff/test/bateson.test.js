/* eslint-disable id-length */
import bateson, {isSameType} from '../src/differs/bateson'
import defaultSummarizers from '../src/differs/defaultSummarizers'

describe('when adding field', () => {
  test('should generate add operation', () => {
    const docA = {
      _type: 'book'
    }

    const docB = {
      _type: 'book',
      author: 'J. K. Rowling'
    }

    const result = bateson(docA, docB, {summarizers: defaultSummarizers})
    expect(result).toMatchSnapshot()
  })
})

describe('when removing field', () => {
  test('should generate remove operation', () => {
    const docA = {
      _type: 'book',
      author: 'J. R. R. Tolkien'
    }

    const docB = {
      _type: 'book'
    }

    const result = bateson(docA, docB, {summarizers: defaultSummarizers})
    expect(result).toMatchSnapshot()
  })
})

describe('when modifying strings', () => {
  test('given string is edited, should generate editText operation', () => {
    const docA = {
      _type: 'book',
      author: 'J. R. R. Tolkien'
    }

    const docB = {
      _type: 'book',
      author: 'J. K. Rowling'
    }

    const result = bateson(docA, docB, {summarizers: defaultSummarizers})
    expect(result).toMatchSnapshot()
  })
})

describe('bateson tests', () => {
  test('image test', () => {
    const docA = {
      _type: 'zoo',
      profile: {
        _type: 'image',
        asset: {
          _type: 'asset',
          _ref: 'ref-123'
        },
        source: 'my-fancy-source'
      }
    }

    const docB = {
      _type: 'zoo',
      profile: {
        _type: 'image',
        asset: {
          _type: 'asset',
          _ref: 'ref-456'
        },
        source: 'my-fancy-source-2'
      }
    }

    const summarizers = {
      image: {
        resolve: (a, b) => {
          if (a.asset && b.asset && a.asset._ref !== b.asset._ref) {
            return [{op: 'my-custom-action', from: a.asset._ref, to: b.asset._ref}]
          }
          // TODO: Call SUMMARIZE()?
          return null
        }
      }
    }

    const result = bateson(docA, docB, {
      summarizers: {...Object.entries(defaultSummarizers), ...summarizers}
    })
    expect(result).toMatchSnapshot()
  })
})

describe('when diffing', () => {
  const zooA = {
    _type: 'zoo',
    keeper: {
      _type: 'keeper',
      name: 'Steve Irwin'
    }
  }

  const zooB = {
    _type: 'zoo',
    keeper: {
      _type: 'keeper',
      name: 'Bindi Irwin ðŸ˜­'
    }
  }

  test('given no summarizer for object, should use default summarizers', () => {
    const result = bateson(zooA, zooB, {summarizers: defaultSummarizers})
    expect(result).toMatchSnapshot()
  })

  test('given summarizer for object, should use custom summarizer', () => {
    const summarizers = {
      keeper: {
        resolve: (a, b) => {
          return [{op: 'keeperNameChanged', from: a.name, to: b.name}]
        }
      }
    }
    const result = bateson(zooA, zooB, {
      summarizers: {...Object.entries(defaultSummarizers), ...summarizers}
    })
    expect(result).toMatchSnapshot()
  })
})

describe('bateson tests', () => {
  test('bateson stuff', () => {
    const zooA = {
      _type: 'zoo',
      keeper: {
        _type: 'keeper',
        face: {
          _type: 'face',
          nose: 'slim'
        }
      },
      zebra: {
        _type: 'zebra',
        face: {
          _type: 'face',
          nose: 'long',
          eyes: 2
        }
      }
    }

    const zooB = {
      _type: 'zoo',
      keeper: {
        _type: 'keeper',
        face: {
          _type: 'face',
          nose: 'big'
        }
      },
      zebra: {
        _type: 'zebra',
        face: {
          _type: 'face',
          nose: 'longandbig',
          eyes: 3
        }
      }
    }

    // TODO: When zebra is defined as a summarizer, it "takes over" the processing of
    //  everything deeper in the structure. By introducing `fields` (as below), bateson can
    //  keep track of what has been processed already and not. We could also potentially
    //  make 'zebra.face.nose' take prescedence over 'face.nose' as 'zebra.face.nose'
    //  is more specific.
    const summarizers = {
      zebra: {
        resolve: (previous, current, path) => {
          return [{op: 'only-zebra-stuff', from: previous, to: current}]
        },
        fields: ['zebra.face.nose']
      },
      face: {
        resolve: (previous, current, path) => {
          return [{op: 'wut', from: previous, to: current}]
        },
        fields: ['face.nose']
      }
    }

    const result = bateson(zooA, zooB, {summarizers})
    expect(result).toMatchSnapshot()
  })
})

describe('when diffing ignored fields', () => {
  test('should not generate diff entries', () => {
    const zooA = {
      _type: 'zoo',
      _id: '123',
      _updatedAt: '2019-02-02T00:00:00.000Z',
      _createdAt: '2019-02-01T00:00:00.000Z',
      _rev: '123',
      _weak: false
    }

    const zooB = {
      _type: 'zoo',
      _id: '456',
      _updatedAt: '2019-02-04T00:00:00.000Z',
      _createdAt: '2019-02-03T00:00:00.000Z',
      _rev: '456',
      _weak: true
    }

    const result = bateson(zooA, zooB, {summarizers: defaultSummarizers})
    expect(result).toEqual([])
  })
})

describe('when checking if two objects are of same Sanity type', () => {
  test.each([
    [null, null, true],
    [null, {_type: 'zoo'}, false],
    [{_type: 'zoo'}, null, false],
    [{}, {}, true],
    [[], [], true],
    [{_type: 'zoo'}, {_type: 'zoo'}, true]
  ])('isSameType(%o, %o)', (a, b, expected) => {
    const result = isSameType(a, b)
    expect(result).toBe(expected)
  })
})

describe('when diffing a document', () => {
  test('should generate a flat array of sumaries', () => {
    const zooA = {
      _id: '123',
      _type: 'zoo',
      keeper: {
        age: 5,
        name: 'Alfred'
      },
      name: 'El Zooas Thomaxas'
    }

    const zooB = {
      _id: 'drafts.123',
      _type: 'zoo',
      keeper: {
        age: 6,
        name: 'Alice'
      },
      name: 'El Zooas Thomaxas'
    }

    const expectedNumberChange = {op: 'edit', from: 5, to: 6}
    const expectedTextChange = {op: 'editText', type: 'string', from: 'Alfred', to: 'Alice'}

    const result = bateson(zooA, zooB, {summarizers: defaultSummarizers})
    expect(result.length).toEqual(2)
    expect(result[0]).toEqual(expectedNumberChange)
    expect(result[1]).toEqual(expectedTextChange)
  })
})

describe('when an object has been added to an array', () => {
  test('should generate a "set" operation', () => {
    const zooA = {
      _id: '123',
      _type: 'zoo',
      openingHours: []
    }

    const zooB = {
      _id: 'drafts.123',
      _type: 'zoo',
      openingHours: [
        {
          _key: '7c842c8c8f38',
          _type: 'openDayAndTime',
          day: 'Tuesday',
          opensAt: '2020-03-03T11:45:22.336Z',
          closesAt: '2020-03-03T12:00:24.349Z'
        }
      ]
    }

    const result = bateson(zooA, zooB, {summarizers: defaultSummarizers})
    expect(result.length).toEqual(1)
    expect(result[0].op).toEqual('set')
  })
})
