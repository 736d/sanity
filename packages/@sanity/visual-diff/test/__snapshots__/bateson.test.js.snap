// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bateson tests bateson stuff 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "changes": Array [
          Object {
            "from": Object {
              "_type": "face",
              "nose": "slim",
            },
            "op": "wut",
            "to": Object {
              "_type": "face",
              "nose": "big",
            },
          },
        ],
        "field": "face",
        "op": "modifyField",
        "type": "face",
      },
    ],
    "field": "keeper",
    "op": "modifyField",
    "type": "keeper",
  },
  Object {
    "changes": Array [
      Object {
        "from": Object {
          "_type": "zebra",
          "face": Object {
            "_type": "face",
            "eyes": 2,
            "nose": "long",
          },
        },
        "op": "only-zebra-stuff",
        "to": Object {
          "_type": "zebra",
          "face": Object {
            "_type": "face",
            "eyes": 3,
            "nose": "longandbig",
          },
        },
      },
    ],
    "field": "zebra",
    "op": "modifyField",
    "type": "zebra",
  },
]
`;

exports[`bateson tests image test 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "from": "ref-123",
        "op": "my-custom-action",
        "to": "ref-456",
      },
    ],
    "field": "profile",
    "op": "modifyField",
    "type": "image",
  },
]
`;

exports[`when adding field should generate add operation 1`] = `
Array [
  Object {
    "field": "author",
    "op": "add",
    "value": "J. K. Rowling",
  },
]
`;

exports[`when diffing given no summarizer for object, should use default summarizers 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "changes": Array [
          Object {
            "from": "Steve Irwin",
            "op": "editText",
            "to": "Bindi Irwin ðŸ˜­",
            "type": "string",
          },
        ],
        "field": "name",
        "op": "modifyField",
        "type": "string",
      },
    ],
    "field": "keeper",
    "op": "modifyField",
    "type": "keeper",
  },
]
`;

exports[`when diffing given summarizer for object, should use custom summarizer 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "from": "Steve Irwin",
        "op": "keeperNameChanged",
        "to": "Bindi Irwin ðŸ˜­",
      },
    ],
    "field": "keeper",
    "op": "modifyField",
    "type": "keeper",
  },
]
`;

exports[`when modifying strings given string is edited, should generate editText operation 1`] = `
Array [
  Object {
    "changes": Array [
      Object {
        "from": "J. R. R. Tolkien",
        "op": "editText",
        "to": "J. K. Rowling",
        "type": "string",
      },
    ],
    "field": "author",
    "op": "modifyField",
    "type": "string",
  },
]
`;

exports[`when removing field should generate remove operation 1`] = `
Array [
  Object {
    "field": "author",
    "op": "remove",
  },
]
`;
